service/example_package_test.go

```diff
@@ -0,0 +1,70 @@
+package service
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/nats-io/nats.go"
+)
+
+func Example() {
+	s := RunServerOnPort(-1)
+	defer s.Shutdown()
+
+	nc, err := nats.Connect(s.ClientURL())
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer nc.Close()
+
+	// Service handler is a function which takes *service.Request as argument.
+	// req.Respond or req.Error should be used to respond to the request.
+	incrementHandler := func(req *Request) {
+		val, err := strconv.Atoi(string(req.Data))
+		if err != nil {
+			req.Error("400", "request data should be a number")
+			return
+		}
+
+		responseData := val + 1
+		req.Respond([]byte(strconv.Itoa(responseData)))
+	}
+
+	config := Config{
+		Name:        "IncrementService",
+		Version:     "v0.1.0",
+		Description: "Increment numbers",
+		Endpoint: Endpoint{
+			// service handler
+			Handler: incrementHandler,
+			// a unique subject serving as a service endpoint
+			Subject: "numbers.increment",
+		},
+	}
+	// Multiple instances of the servcice with the same name can be created.
+	// Requests to a service with the same name will be load-balanced.
+	for i := 0; i < 5; i++ {
+		svc, err := Add(nc, config)
```

Comment:
It's in separate package now, so the way you actually use it is by calling `service.Add()`, which I think is really intuitive and pretty much a standard in Go (e.g. `list.New()` or `context.WithTimeout()`, `http.Server()` etc.).

In that sense, `service.AddService()` seems redundant.

Reply:
I thought you would hang it off alive connection. nc.AddService. This is how we register interest with a connection, and that is what a service it imo. Do you now make them pass in a nats.Connection?

Reply:
Yes - I want to separate services to a separate, importable package - if user does not want to use service framework, the API is less cluttered and they do not see methods they do not need. After importing `service` package and typing `service.` in IDE, you'll see all you need to start using interfaces. It's self-documenting and easier to use.

Reply:
We should discuss, we had stated we did not want this as a separate package IIRC because we wanted this to be the default way to do micro-services etc.

Let's schedule some time to discuss before this gets merged.

Reply:
Example of adding a new service looks sth like this:
```go
	nc, err := nats.Connect("127.0.0.1:4222")
	if err != nil {
		log.Fatal(err)
	}

	config := Config{
		Name:     "EchoService",
		Endpoint: Endpoint{Subject: "echo", Handler: echoHandler},
	}

	srv, err := service.Add(nc, config)
	if err != nil {
		log.Fatal(err)
	}
```

Reply:
Sure, let's discuss this

Reply:
The part where `nc.AddService` may come in handy is when you have multiple services on top of the same connection and you want to be able to use the error handling on a per service basis, since `nc.Opts.AsyncErrorCB` belongs to the connection, in current PR it seems that only done Service would be able to use the error handler.

Reply:
Current service implementation also has `ErrorHandler`, which extends `AsyncErrorCB`.

Reply:
This is the kind of package that will sprawl with dependencies like observability and more over time.  Might be worth keeping in a seperate package to avoid those being in nats.go core

Reply:
@ripienaar You have a good point here, but unfortunately, separate packages have the same `go.mod`.

We would need a multi-module setup. That brings it's own problems and complexity (especially with incompatible versions of dependencies).